File: 102-levelOrder.java
/*
It is important to master List<List<Integer>> related commands and recursive usage.
*/

import java.util.ArrayList;
import java.util.List;

public class Main {

    public static void main(String[] args) {
        TreeNode A1 = new TreeNode(3);
        TreeNode A2 = new TreeNode(9);
        TreeNode A3 = new TreeNode(20);
        TreeNode A4 = new TreeNode(15);
        TreeNode A5 = new TreeNode(7);

        A1.left = A2;
        A1.right = A3;
        A2.left = null;
        A2.right = null;
        A3.left = A4;
        A3.right = A5;
        A4.left = null;
        A4.right = null;
        A5.left = null;
        A5.right = null;

        List<List<Integer>> result = levelOrder(A1);
        System.out.println(result);
    }


    public static List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        levelRecursion(root,result,0);
        return result;
    }

    private static void levelRecursion(TreeNode node, List<List<Integer>> result, int level){
        if (node==null){
            return;
        }
        if (result.size()<level+1){
            result.add(new ArrayList<Integer>());
        }
        result.get(level).add(node.val);
        levelRecursion(node.left, result, level+1);
        levelRecursion(node.right, result, level+1);
    }

}

File: 102-levelOrder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if root is None:
            return []
        result = []
        stack = [root]
        while(len(stack) > 0):
            result.append([s.val for s in stack])
            stack_num = len(stack)
            for i in range(stack_num):
                node = stack.pop(0)
                if node.left is not None:
                    stack.append(node.left)
                if node.right is not None:
                    stack.append(node.right)
        return result
            
        